app/src/main/java/com/tatar/currencify/presentation/feature/CurrencyConversionViewModel.kt

```diff
@@ -0,0 +1,120 @@
+
+    private val _currencyConversionResult = MutableLiveData<CurrencyConversionResult>()
+    val currencyConversionResult: LiveData<CurrencyConversionResult> get() = _currencyConversionResult
+
+    private val _loadingState = MutableLiveData<Boolean>()
+    val loadingState: LiveData<Boolean> get() = _loadingState
+
+    private val _responderClickEvent = MutableLiveData<Event<Unit>>()
+    val responderClickEvent: LiveData<Event<Unit>> get() = _responderClickEvent
+
+    private val _animationEvent = MutableLiveData<Event<Unit>>()
+    val animationEvent: LiveData<Event<Unit>> get() = _animationEvent
+
+    private val _hideKeyboardEvent = MutableLiveData<Event<Unit>>()
+    val hideKeyboardEvent: LiveData<Event<Unit>> get() = _hideKeyboardEvent
+
+    private val _errorEvent = MutableLiveData<Event<Unit>>()
+    val errorEvent: LiveData<Event<Unit>> get() = _errorEvent
+
+    private val amountSubject: BehaviorSubject<Double> = BehaviorSubject.createDefault(INITIAL_BASE_CURRENCY_AMOUNT)
+
+    init {
+        startCurrencyTransformationFlow(INITIAL_BASE_CURRENCY)
+    }
+
+    fun onCurrencyConversionItemClick(baseCurrency: String) {
+        hideKeyboard()
+        stopCurrencyTransformationFlow()
+        startCurrencyTransformationFlow(baseCurrency)
+    }
+
+    private fun hideKeyboard() {
+        _hideKeyboardEvent.value = (Event(Unit))
+    }
+
+    private fun stopCurrencyTransformationFlow() {
+        if (disposables.size() > 0) disposables.clear()
+    }
+
+    private fun startCurrencyTransformationFlow(baseCurrency: String) {
+        Flowables.combineLatest(
+            getRatesTask.invoke(baseCurrency),
+            amountSubject.toFlowable(BackpressureStrategy.LATEST) // TODO use flowable directly maybe? correct strategy?
+        ) { latestRatesEntity, amount -> mapToCurrencyConversionResult(latestRatesEntity, amount) }
+            .doOnSubscribe { onNewFlowRequest() }
+            .subscribe(
+                { currencyConversionResult -> onSuccess(currencyConversionResult) },
+                { onError() }
+            ).addTo(disposables)
+    }
+
+    private fun mapToCurrencyConversionResult(
+        latestRatesEntity: LatestRatesEntity,
+        amount: Double
+    ): CurrencyConversionResult {
+        return CurrencyConversionResult(
+            baseCurrency = latestRatesEntity.baseCurrency,
+            convertedCurrencies = latestRatesEntity.rates.map {
+                ConvertedCurrency(it.currencyCode, it.rate * amount)
+            }
+        )
+    }
+
+    private fun onNewFlowRequest() = _loadingState.postValue(true)
+
+    private fun onSuccess(currencyConversionResult: CurrencyConversionResult) {
+        if (isNewFlow(loadingState.value)) {
+            _loadingState.postValue(false)
+            _animationEvent.postValue(Event(Unit))
+        }
+
+        _currencyConversionResult.postValue(currencyConversionResult)
+    }
+
+    private fun isNewFlow(isLoading: Boolean?) = isLoading == true
```

Comment:
`Boolean?` means what? Does it nullable?

Reply:
Yup

Reply:
isLoading can be nil you mean, how? It should be either loading or notLoading.

Reply:
`isLoading` value comes from the `_loadingState` member variable and `_loadingState` is `LiveData` field. You can think of it some kind of an observable generic data holder. Initially `LiveData` values are set to null.